Init Generator layer.

dereuromark opened the pull request:
Resolves
https://github.com/vierge-noire/cakephp-fixture-factories/issues/263

Not sure about the whole overhead of optional vs unique
Got a bit out of hand trying to support both.

Open questions:
- How to clarify new methods on DummyData that dont exist on Faker? the annotations then display it even if not existing on old one?
- How to shim e.g. methods that work now fine on DummyData, but would require shimming on old one to work the same when switching the generator type

Open topics:
- Fixing tests (PHPStan, CS is green, but tests need further fixing)

Requires PHP8.2 for random generator topics.

Also:
- updated migrations to use new Base classes (future compatibility of new major soon)
- phpstan higher level now with some code cleanup

dereuromark committed:
Init Generator layer.

dereuromark committed:
Init Generator layer.

dereuromark committed:
Fix up phpstan level 6.

dereuromark committed:
Fix up phpstan level 6.

johnykvsky commented on src/Generator/DummyGeneratorAdapter.php:
```diff
@@ -0,0 +1,194 @@
+     */
+    public function seed(?int $seed = null): void
+    {
+        if ($seed !== null) {
```
```
    $this->generator->addDefinition(
        \DummyGenerator\Definitions\Randomizer\RandomizerInterface::class,
        new \DummyGenerator\Core\Randomizer\XoshiroRandomizer(seed: $seed)
    );
```

It will replace default randomizer with the one that supports seed.

dereuromark replied:
Can you elaborate on your plan?

Apparently, I asked AI as well, there is an issue.

johnykvsky replied:
I have updated docs for it, please look [here](https://github.com/johnykvsky/dummygenerator/blob/main/docs/howto.md#how-can-i-use-seed)

So it should work with:
```
    public function seed(?int $seed = null): void
    {
      $this->generator->addDefinition(
        \DummyGenerator\Definitions\Randomizer\RandomizerInterface::class,
        new \DummyGenerator\Core\Randomizer\XoshiroRandomizer(seed: $seed)
       );
   }
```

and it will make generator work with given seed.

johnykvsky replied:
Well, maybe one small update: you need to check if seed is not null before going to `$this->generator->addDefinition(...`.

johnykvsky replied:
Full working script taken from [this](https://github.com/FakerPHP/Faker/issues/870) issue:

```
<?php

declare(strict_types=1);

require_once __DIR__ . '/vendor/autoload.php';

$pn1 = null;
$pn2 = null;

for ($i = 0; $i < 3; $i++) {
    // $faker = Faker\Factory::create('en_US');
    // $faker->seed(1234567);

    $container = \DummyGenerator\Container\DefinitionContainerBuilder::base(); 
    $faker = new \DummyGenerator\DummyGenerator($container);  
    $faker->addDefinition(
        \DummyGenerator\Definitions\Randomizer\RandomizerInterface::class,
        new \DummyGenerator\Core\Randomizer\XoshiroRandomizer(seed: 123)
    );

    $n1 = $faker->numberBetween(1000, 10000);
    if ($pn1 !== null && $pn1 !== $n1) {
        echo sprintf('DELTA in N1: previous: %d, this time: %d, iteration: %d', $pn1, $n1, $i) . PHP_EOL;
    }
    $pn1 = $n1;

    gc_collect_cycles(); // force GC run

    $n2 = $faker->numberBetween(1000, 10000);
    if ($pn2 !== null && $pn2 !== $n2) {
        echo sprintf('DELTA in N2: previous: %d, this time: %d, iteration: %d', $pn2, $n2, $i) . PHP_EOL;
    }
    $pn2 = $n2;

    var_dump([$n1, $n2]);
}
```

johnykvsky commented on src/Generator/DummyUniqueAdapter.php:
```diff
@@ -0,0 +1,84 @@
+     */
+    public function seed(?int $seed = null): void
+    {
+        $this->generator->seed($seed);
```
Copy/paste from Faker? This will not work with `DummyGenerator`

johnykvsky commented on src/Generator/DummyGeneratorAdapter.php:
```diff
@@ -0,0 +1,194 @@
+
+        // DummyGenerator uses __call for all its methods, so we try to call it directly
+        try {
+            if ($this->isUnique) {
```
To get unique value with `DummyGenerator`:

```
$generator = new DummyGenerator(DefinitionContainerBuilder::all(), new UniqueStrategy(retries: 1000));
```

johnykvsky replied:
Or you can do that like this:

```
$this->geneator->withStrategy(new UniqueStrategy(1000))->firstName();
```

It will keep `$this->generator` as it was, but execute `firstName()` method on generator with unique strategy and max repeat = 1000

dereuromark committed:
Try to include review comments.

dereuromark commented:
1. Implemented UniqueStrategy configuration: The DummyGeneratorAdapter now uses UniqueStrategy(retries: 1000) as suggested by reviewer johnykvsky.
  2. Added note about seeded randomizer: While the reviewer suggested using XoshiroRandomizer with a seed, the current version of DummyGenerator (0.0.5) doesn't support this pattern directly. I've added a comment explaining
   this limitation in the seed() method.

dereuromark committed:
Try to include review comments.

dereuromark committed:
Add enum shim.

dereuromark commented:
I wonder why the enum() method for example isnt directy available in the factory here
```php
$this->setDefaultData(function (GeneratorInterface $generator) {
     return [
        'first_name' => $generator->firstName(),
        'last_name' => $generator->lastName(),
        'phone' => implode(" ", $generator->words(3)),
        'type' => $generator->enum(TeamMemberType::class),
    ];
```
I shimmed it for now, that seems to work

johnykvsky commented:
> I wonder why the enum() method for example isnt directy available in the factory here
> 
> ```
> $this->setDefaultData(function (GeneratorInterface $generator) {
>      return [
>         'first_name' => $generator->firstName(),
>         'last_name' => $generator->lastName(),
>         'phone' => implode(" ", $generator->words(3)),
>         'type' => $generator->enum(TeamMemberType::class),
>     ];
> ```
> 
> I shimmed it for now, that seems to work

How do you initialize `DummyGenerator`? It should work and be available with `DefinitionContainerBuilder::base()` and above (like `::default()`, `::all()`. It's not in the `::core()`

dereuromark commented:
So enumElement() returns the object, and enumValue() the scalar value, got it.
Makes sense to me. Thx!

@LordSimal 
For me it works with both approaches with a much cleaner API now (methods only), so the dropdown is more concise:
<img width="642" height="374" alt="generator" src="https://github.com/user-attachments/assets/fefa3977-d3b6-4d7a-8da8-7751daf6db00" />

@ others:
Any feedback so far on this?

dereuromark committed:
Fix up latest enum changes.

dereuromark committed:
Fix up latest enum changes.

dereuromark committed:
Fix tests.

dereuromark committed:
Fix tests.

dereuromark committed:
Fix tests.

dereuromark committed:
Fix Psalm.
